Libraries/VtkVgCore/vtkVgSGIReader.cxx

```diff
@@ -0,0 +1,358 @@
+void DiskToHardware(uint32_t& val) { val = ntohl(val); }
+
+//-----------------------------------------------------------------------------
+template <typename Enum>
+void DiskToHardware(Enum& enumValue)
+{
+  using Underlying = typename std::underlying_type<Enum>::type;
+  auto& castedValue = reinterpret_cast<Underlying&>(enumValue);
+  DiskToHardware(castedValue);
+}
+
+} // namespace <anonymous>
+
+//-----------------------------------------------------------------------------
+class vtkVgSGIReader::vtkInternal
+{
+public:
+  vtkSmartPointer<vtkImageData> ImageData;
+};
+
+//-----------------------------------------------------------------------------
+vtkVgSGIReader::vtkVgSGIReader() : Internal{new vtkInternal}
+{
+  this->SetNumberOfInputPorts(0);
+  this->SetNumberOfOutputPorts(1);
+}
+
+//-----------------------------------------------------------------------------
+vtkVgSGIReader::~vtkVgSGIReader()
+{
+  delete this->Internal;
+}
+
+//-----------------------------------------------------------------------------
+void vtkVgSGIReader::PrintSelf(ostream& os, vtkIndent indent)
+{
+  this->Superclass::PrintSelf(os, indent);
+}
+
+//-----------------------------------------------------------------------------
+int vtkVgSGIReader::GetNumberOfLevels() const
+{
+  return 1;
+}
+
+//-----------------------------------------------------------------------------
+void vtkVgSGIReader::GetDimensions(int dim[2])
+{
+  if (this->Internal->ImageData)
+    {
+    int fulldim[3];
+    this->Internal->ImageData->GetDimensions(fulldim);
+    dim[0] = fulldim[0];
+    dim[1] = fulldim[1];
+    }
+  else
+    {
+    dim[0] = dim[1] = 0;
+    }
+}
+
+//----------------------------------------------------------------------------
+bool vtkVgSGIReader::CanRead(const std::string& source)
+{
+  std::string ciSource = vtksys::SystemTools::LowerCase(source);
+
+  // Check if it is a filename or if the argument is just an extension.
+  std::string ext = vtksys::SystemTools::GetFilenameLastExtension(ciSource);
+  if (ext.compare(".sgi") == 0 || (ciSource.compare("sgi") == 0) ||
+      (ciSource.compare(".sgi") == 0))
+    {
+    return true;
+    }
+
+  return false;
+}
+
+//----------------------------------------------------------------------------
+std::string vtkVgSGIReader::GetShortDescription()
+{
+  return "Reader for SGI file format";
+}
+
+//----------------------------------------------------------------------------
+std::string vtkVgSGIReader::GetLongDescription()
+{
+  return this->GetShortDescription();
+}
+
+//----------------------------------------------------------------------------
+vtkVgBaseImageSource* vtkVgSGIReader::Create()
+{
+  return vtkVgSGIReader::New();
+}
+
+//-----------------------------------------------------------------------------
+int vtkVgSGIReader::RequestInformation(
+  vtkInformation* vtkNotUsed(request),
+  vtkInformationVector** vtkNotUsed(inputVector),
+  vtkInformationVector* outputVector)
```

Comment:
~~Can this method be called more than once? Should it re-read the file each time, or cache it after the first?~~ Redacted after examining the implementation of `vtkVgPNGReader` and seeing the similarity.